lib/astc_encode.cpp

```diff
@@ -0,0 +1,925 @@
+                input_image = unorm8x2ArrayToImage(This->pData + offset,
+                                                   width, height);
+            else if (num_components == 3)
+                input_image = unorm8x3ArrayToImage(This->pData + offset,
+                                                   width, height);
+            else // assume (num_components == 4)
+                input_image = unorm8x4ArrayToImage(This->pData + offset,
+                                                   width, height);
+
+            assert(input_image);
+
+            CompressionWorkload work;
+            work.context = astc_context;
+            work.image = input_image;
+            work.swizzle = swizzle;
+            work.data_out = buffer_out;
+            work.data_len = levelImageSizeOut;
+            work.error = ASTCENC_SUCCESS;
+
+            launchThreads(threadCount, compressionWorkloadRunner, &work);
+
+            if (work.error != ASTCENC_SUCCESS) {
+                std::cout << "ASTC compressor failed\n" <<
+                             astcenc_get_error_string(work.error) << std::endl;
+
+                imageFree(input_image);
+
+                astcenc_context_free(astc_context);
+                return KTX_INVALID_OPERATION;
+            }
+
+            buffer_out += levelImageSizeOut;
+
+            // Reset ASTC context for next image
+            astcenc_compress_reset(astc_context);
+            offset += levelImageSizeIn;
+        }
+    }
+
+    // We are done with astcencoder
+    astcenc_context_free(astc_context);
+
+    // reset pointer to new BDB
+    BDB = prototype->pDfd+1;
+
+    // ASTC-related checks
+    ktx_uint32_t model = KHR_DFDVAL(BDB, MODEL);
+    ktx_uint32_t primaries = KHR_DFDVAL(BDB, PRIMARIES);
+
+    assert(model == KHR_DF_MODEL_ASTC && "Invalid dfd generated for ASTC image\n");
+    if (transfer == KHR_DF_TRANSFER_SRGB) {
+        assert(primaries == KHR_DF_PRIMARIES_SRGB && "Not a valid sRGB image\n");
+    }
+
+    // Fix up the current (This) texture
+    #undef DECLARE_PRIVATE
+    #undef DECLARE_PROTECTED
+    #define DECLARE_PRIVATE(n,t2) ktxTexture2_private& n = *(t2->_private)
+    #define DECLARE_PROTECTED(n,t2) ktxTexture_protected& n = *(t2->_protected)
+
+    DECLARE_PROTECTED(thisPrtctd, This);
+    DECLARE_PRIVATE(protoPriv, prototype);
+    DECLARE_PROTECTED(protoPrtctd, prototype);
+    memcpy(&thisPrtctd._formatSize, &protoPrtctd._formatSize,
+           sizeof(ktxFormatSize));
+    This->vkFormat = vkFormat;
+    This->isCompressed = prototype->isCompressed;
+    This->supercompressionScheme = KTX_SS_NONE;
+    This->_private->_requiredLevelAlignment = protoPriv._requiredLevelAlignment;
+    // Copy the levelIndex from the prototype to This.
+    memcpy(This->_private->_levelIndex, protoPriv._levelIndex,
+           This->numLevels * sizeof(ktxLevelIndexEntry));
+    // Move the DFD and data from the prototype to This.
+    delete This->pDfd;
```

Comment:
This and the `delete` below needs to be `free(..)` not `delete` because the memory was allocated by `malloc`. I can't remember the details now but I was doing the same thing in `ktxTexture2_BasisEncodeEx` and somebody reported errors with a certain compiler.

Reply:
Agree. This is also copy pasted from `ktxTexture2_TranscodeBasis` so needs changing there too.

Reply:
Oops. I thought I'd fixed it there when the bug was reported. Please fix it.

Reply:
Fixed in https://github.com/KhronosGroup/KTX-Software/pull/433/commits/18af4983456764eb784babe2d11a125f5d2b7f5c